nexus/src/app/sagas/instance_update/mod.rs

```diff
@@ -0,0 +1,619 @@
+        }
+
+        Ok(builder.build()?)
+    }
+}
+
+async fn siu_become_updater(
+    sagactx: NexusActionContext,
+) -> Result<instance::UpdaterLock, ActionError> {
+    let RealParams {
+        ref serialized_authn, ref authz_instance, orig_lock, ..
+    } = sagactx.saga_params::<RealParams>()?;
+    let saga_id = sagactx.lookup::<Uuid>(INSTANCE_LOCK_ID)?;
+    let opctx =
+        crate::context::op_context_for_saga_action(&sagactx, serialized_authn);
+    let osagactx = sagactx.user_data();
+
+    slog::debug!(
+        osagactx.log(),
+        "instance update: trying to become instance updater...";
+        "instance_id" => %authz_instance.id(),
+        "saga_id" => %saga_id,
+        "parent_lock" => ?orig_lock,
+    );
+
+    let lock = osagactx
+        .datastore()
+        .instance_updater_inherit_lock(
+            &opctx,
+            &authz_instance,
+            orig_lock,
+            saga_id,
+        )
+        .await
+        .map_err(ActionError::action_failed)?;
+
+    slog::info!(
+        osagactx.log(),
+        "Now, I am become Updater, the destroyer of VMMs.";
+        "instance_id" => %authz_instance.id(),
+        "saga_id" => %saga_id,
+    );
+
+    Ok(lock)
+}
+
+async fn siu_unbecome_updater(
+    sagactx: NexusActionContext,
+) -> Result<(), anyhow::Error> {
+    let RealParams { ref serialized_authn, ref authz_instance, .. } =
+        sagactx.saga_params::<RealParams>()?;
+    unlock_instance_inner(serialized_authn, authz_instance, &sagactx, None)
+        .await?;
+
+    Ok(())
+}
+
+async fn siu_update_network_config(
+    sagactx: NexusActionContext,
+) -> Result<(), ActionError> {
+    let Params { ref serialized_authn, ref authz_instance, .. } =
+        sagactx.saga_params()?;
+    let opctx =
+        crate::context::op_context_for_saga_action(&sagactx, serialized_authn);
+    let osagactx = sagactx.user_data();
+    let nexus = osagactx.nexus();
+    let instance_id = InstanceUuid::from_untyped_uuid(authz_instance.id());
+
+    let update =
+        sagactx.lookup::<NetworkConfigUpdate>(NETWORK_CONFIG_UPDATE)?;
+
+    match update {
+        NetworkConfigUpdate::Delete => {
+            info!(
+                osagactx.log(),
+                "instance update: deleting network config";
+                "instance_id" => %instance_id,
+            );
+            nexus
+                .instance_delete_dpd_config(&opctx, authz_instance)
+                .await
+                .map_err(ActionError::action_failed)?;
+        }
+        NetworkConfigUpdate::Update(active_propolis_id) => {
+            // Look up the ID of the sled that the instance now resides on, so that we
+            // can look up its address.
+            let new_sled_id = osagactx
+                .datastore()
+                .vmm_fetch(&opctx, authz_instance, &active_propolis_id)
+                .await
+                .map_err(ActionError::action_failed)?
+                .sled_id;
```

Comment:
A VMM's sled ID is immutable, so I think we can get this from the snapshot instead (assuming we're willing to serialize it into this node's params).

Reply:
Yeah, this was just me copying the preexisting code from
`instance_ensure_network_config`. Good call.

Reply:
I did this in 9fb5aa1cf67f598ad212bc8f41f3e0db5b471683, but had to temporarily revert that change since it uncovered a bit of an issue where we tried to do the migration network config update twice after the migration _should_ have been unlinked. Will continue trying to figure that out.

Reply:
Okay, fixed that up in 730814b86e88e5935b5ca4480d734ec437a41226 and also made the logic for determining how to handle migration updates more correct in general. We now ensure that we only do the network config stuff the _first_ time either side of the migration says it's complete, rather than trying to do it twice. I also fixed an issue where we would accidentally allow new migrations if either side reports complete, rather than only after the target does so.